File: solutions/1448.py
# python 3

class Solution:
    @staticmethod
    def goodNodesImpl(root: TreeNode, max_observed_value: float) -> int:
        if root is None:
            return 0
        retval = int(root.val >= max_observed_value)
        max_observed_value = max(max_observed_value, root.val)
        return retval + sum([
            Solution.goodNodesImpl(child, max_observed_value)
            for child in [root.left, root.right]
        ])

    def goodNodes(self, root: TreeNode) -> int:
        return self.goodNodesImpl(root, float('-inf'))

File: solutions/1448.cpp
#include <limits>

class Solution {
 public:
  int goodNodes(TreeNode* root,
                int max_value_observed = std::numeric_limits<int>::min()) {
    if (not root) {
      return 0;
    }

    int retval = root->val >= max_value_observed;

    for (auto child : {root->left, root->right}) {
      retval += goodNodes(child, max(max_value_observed, root->val));
    }

    return retval;
  }
};
